File: app/src/main/java/kr/co/yogiyo/rookiephotoapp/settings/SignOutDialogFragment.java
package kr.co.yogiyo.rookiephotoapp.settings;

import android.content.Context;
import android.support.v7.app.AlertDialog;
import android.support.v7.preference.Preference;
import android.support.v7.preference.PreferenceDialogFragmentCompat;
import android.view.View;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.google.firebase.auth.FirebaseUser;

import kr.co.yogiyo.rookiephotoapp.R;

public class SignOutDialogFragment extends PreferenceDialogFragmentCompat
        implements View.OnClickListener, SignCallback {

    private Context context;

    public static SignOutDialogFragment newInstance(){
        return new SignOutDialogFragment();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        this.context = context;
    }

    @Override
    protected void onBindDialogView(View view) {
        super.onBindDialogView(view);
        RelativeLayout signoutDialogRelative = view.findViewById(R.id.relative_signout_dialog);
        TextView cancelText = view.findViewById(R.id.text_cancel);
        TextView signoutText = view.findViewById(R.id.text_signout);
        ((SettingsActivity) context).addProgressBarInto(signoutDialogRelative);

        cancelText.setOnClickListener(this);
        signoutText.setOnClickListener(this);
    }

    @Override
    protected void onPrepareDialogBuilder(AlertDialog.Builder builder) {
        super.onPrepareDialogBuilder(builder);
        builder.setTitle("로그아웃")
                .setPositiveButton(null, null)
                .setNegativeButton(null, null);
    }

    @Override
    public void onDialogClosed(boolean positiveResult) {
        // Do nothing
    }

    // TODO : 콜백을 RxJava로 바꿀 수 있을지 고민하기
    // TODO : 취소할 때 요청 취소할 수 있는지 조사
    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.text_cancel:
                dismiss();
                break;
            case R.id.text_signout:
                ((SettingsActivity) context).showLoading();
                ((AuthNavigator) context).signOut(this);
                break;
        }
    }

    @Override
    public void onSuccess(FirebaseUser user) {
        Preference preference = getPreference();
        preference.setTitle(getString(R.string.text_need_to_signin));
        ((SettingsActivity) context).showToast("로그아웃 성공");
        dismiss();
    }

    @Override
    public void onFail() {
        ((SettingsActivity) context).showToast("로그아웃 실패");
        dismiss();
    }
}

File: app/src/main/java/kr/co/yogiyo/rookiephotoapp/settings/SignOutDialogFragment.kt
package kr.co.yogiyo.rookiephotoapp.settings

import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.preference.PreferenceDialogFragmentCompat

import kr.co.yogiyo.rookiephotoapp.R

class SignOutDialogFragment : PreferenceDialogFragmentCompat() {

    lateinit var signOut: () -> Unit

    // TODO : 콜백을 RxJava로 바꿀 수 있을지 고민하기
    override fun onPrepareDialogBuilder(builder: AlertDialog.Builder?) {
        super.onPrepareDialogBuilder(builder)
        builder?.run {
            setTitle(R.string.text_signout)
            setMessage(R.string.text_ask_signout)
            setPositiveButton(R.string.text_dialog_ok, null)
            setNegativeButton(R.string.text_dialog_no) { _, _ -> dismiss() }
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        dialog.setOnShowListener { dialog ->
            (dialog as AlertDialog).getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener {
                (context as SettingsActivity).showLoading(true)
                signOut()
            }
        }
    }

    override fun onDialogClosed(positiveResult: Boolean) {
        // Do nothing
    }

    fun onSuccess() {
        (context as SettingsActivity).run {
            showLoading(false)
            showToast(R.string.text_signout_success)
        }
        dismiss()
    }

    fun onFail() {
        (context as SettingsActivity).run {
            showLoading(false)
            showToast(R.string.text_signout_fail)
        }
        dismiss()
    }

    companion object {

        fun newInstance(): SignOutDialogFragment {
            return SignOutDialogFragment()
        }
    }
}